test/distributed/org/apache/cassandra/distributed/test/ReadRepairQueryTypesTest.java

```diff
@@ -0,0 +1,1215 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.collect.Iterators;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.ICoordinator;
+import org.apache.cassandra.service.reads.repair.ReadRepairStrategy;
+
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.ALL;
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
+import static org.apache.cassandra.utils.ByteBufferUtil.EMPTY_BYTE_BUFFER;
+
+/**
+ * Tests basic read repair functionality with different types of query types and schemas.
+ * <p>
+ * Each test verifies that its tested query triggers read repair propagating mismatched rows/columns and row/column
+ * deletions. They also verify that the selected rows and columns are propagated through read repair on missmatch,
+ * and that unselected rows/columns are not repaired.
+ * <p>
+ * The tests are parameterized for:
+ * <ul>
+ *     <li><Both {@code NONE} and {@code BLOCKING} read repair stratregies/li>
+ *     <li>Data to be repaired residing on the query coordinator or a replica</li>
+ *     <li>Data to be repaired residing on memtables or flushed to sstables</li>
+ * </ul>
+ * <p>
+ * All the included tests have a similar behaviour:
+ * <ul>
+ *     <li>Create a keyspace with RF=2 and a table</li>
+ *     <li>Insert some data in only one of the nodes</li>
+ *     <li>Run the tested read query selecting a subset of the inserted columns with CL=ALL</li>
+ *     <li>Verify that the previous read has triggered read repair propagating only the queried columns</li>
```

Comment:
Does this subsume `ReadRepairTest #alterRFAndRunReadRepair()`? (If not, could if be modified easily to do so?)

Reply:
I think `alterRFAndRunReadRepair` tests both RR partial propagation of columns and altering RF. I have kept it mainly because of the alter RF part, and because I'm always afraid of removing tests in case the new replacement tests could be inadvertently missing something from the original scenario.

Reply:
The test around altering RF seems useful to me. I would keep it.